kong/core/router.lua

```diff
@@ -600,8 +601,22 @@ function _M.new(apis)
     end
 
 
-    if api_t.upstream.path then
-      new_uri = api_t.upstream.path .. new_uri
+    local upstream_path = api_t.upstream.path
+    if upstream_path then
+      if new_uri == "/" then
+        new_uri = upstream_path
+      else
+        new_uri = upstream_path .. (sub(upstream_path, -1) == "/" and sub(new_uri, 2) or new_uri)
+      end
+    end
+
+
+    local req_uri_slash = sub(uri,     -1) == "/"
+    local new_uri_slash = sub(new_uri, -1) == "/"
+    if new_uri_slash and not req_uri_slash and new_uri ~= "/" then
```

Comment:
@Tieske I once asked @coopr, and he said that we don't intently have boy scout practices in place. And I believe I will end down to a rabbit hole with this patch, if I try. As the style guide is for me mostly a fuzzy logic with many many exceptions like the above. Even the above comment makes me think, what is "complex".

E.g. is a function call with many arguments a complex?

```lua
local api_t = find_api(method, uri, req_host)
```

What about:

```
local api_t = find_api(method, uri, {
    something = 1
})
```

 (is that still simple?)